### PR TITLE
incremental restyle: Use more concrete types in the style system

### PR DESCRIPTION
**Pull request by bholley**
We need to hang both snapshots and restyle damage off of ElementData, and so we need them to be concrete to avoid infecting ElementData with the trait hierarchy.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14034)

### CODE REVIEW

File: components/style/dom.rs [hunk 1]
```diff
@@ -186,15 +184,15 @@ pub trait TElement : PartialEq + Debug + Sized + Copy + Clone + ElementExt + Pre
...
+    fn set_restyle_damage(self, damage: RestyleDamage);
 
     /// XXX: It's a bit unfortunate we need to pass the current computed values
     /// as an argument here, but otherwise Servo would crash due to double
     /// borrows to return it.
     fn existing_style_for_restyle_damage<'a>(&'a self,
                                              current_computed_values: Option<&'a Arc<ComputedValues>>,
                                              pseudo: Option<&PseudoElement>)
-        -> Option<&'a <Self::ConcreteRestyleDamage as TRestyleDamage>::PreExistingComputedValues>;
+        -> Option<&'a <RestyleDamage as TRestyleDamage>::PreExistingComputedValues>;
```
Comment (emilio) [inline components/style/dom.rs]: You might be able to leave this as `RestyleDamage::PreExistingComputedValues`.

**Review by emilio — APPROVED** [review]
r=me

File: components/style/dom.rs (hunk 1 above)
Comment (bholley) [inline components/style/dom.rs]: Nope, it's an ambiguous associated type unfortunately.

**Comment by nox** [comment]
This needs a rebase.

**Comment by KiChjang** [comment]
@bors-servo r-

Removing from homu queue until it's rebased.

**Comment by bholley** [comment]
Bad rebase.

@bors-servo r=emilio

**Comment by jdm** [comment]
@bors-servo: retry
* #14118

**Pull request merged**